scanner/decisions/0002-local-scanning-with-scanner-v4.md

```diff
@@ -0,0 +1,64 @@
+
+Currently, Sensor and Central coordinate the calls to Scanner and Scanner Slim. The local image scanning leverages `GetImageComponents()` in Scanner V2 Slim and `GetImageVulnerabilities()` in Central Scanner V2. Both APIs have similar semantics to Scanner V4's Indexer and Matcher services:
+
+| Scanner V2              | Scanner V4                 | Description                                                           |
+|-------------------------|----------------------------|-----------------------------------------------------------------------|
+| GetImageComponents      | Indexer/CreateIndexReport  | Retrieve the inventory of artifacts and details on the image content. |
+| GetImageVulnerabilities | Matcher/GetVulnerabilities | Retrieve the matching vulnerabilities for the components provided.    |
+
+Although `GetImageVulnerabilities` and `Matcher/GetVulnerabilities` are similar, they have a crucial difference. `Matcher/GetVulnerabilities` was specifically designed to retrieve the index report to be scanned from a separate service or storage, similar to how [Clair is implemented](https://github.com/quay/clair/blob/main/httptransport/matcher_v1.go#L116). On the other hand, `GetImageVulnerabilities` accepts a list of components in its payload. To fully utilize the clear separation between Indexer and Matcher in Scanner V4's design, changes need to be made to Scanner V4, Central, and Sensor. This includes Central and Sensor's ability to handle Index Reports and also Scanner V4's capability to access reports that are generated in the secured cluster.
+
+## Decision
+
+Scanner V4's Matcher will accept a new parameter in `GetVulnerabilitiesRequest` to support local scanning. This parameter will allow clients to provide the `v4.IndexReport` for scanning. The Matcher will then verify if the report components are available in the request. If not, it will retrieve them from the `Indexer`.
+
+Example:
+
+```proto
+message GetVulnerabilitiesRequest {
+        string hash_id = 1;
+        Contents contents = 2;
+}
+```
+
+Notice that we don't necessarily need to embed the whole Index Report since the Matcher only needs a subset of the proto fields:
+
+```proto
+message Contents {
+    repeated Package              packages      = 1;
+    repeated Distribution         distributions = 2;
+    repeated Repository           repositories  = 3;
+    map<string, Environment.List> environments  = 4;
+}
+```
+
+But that will be left to the implementation to decide.
+
+Sensor and Central will behave the same as today regarding handling scan requests. Scanner V4 clients will have to make minimal changes to the interfaces of current scanner clients[^1] to support both V2 and V4. In central, a new field for the `IndexreportComponents` will be added [to the `EnrichLocalImageInternalRequest`](https://github.com/stackrox/stackrox/blob/a21793de1842586499e4afb3de68b780753db7f0/proto/api/v1/image_service.proto#L62).
```

Comment:
Thoughts on creating new Central APIs for the V4 use cases? (ie: create a new endpoint instead of re-using `EnrichLocalImageInternal`).

The current `EnrichLocalImageInternal` has some debt, by separating the APIs it may be easier to retire the old API in the future when support lapses.

Reply:
> Thoughts on creating new Central APIs for the V4 use cases?

The thinking behind reusing is to reduce the scope of the changes to introduce scanner v4. A new API would probably need to do the same as the old one, and although I agree the old implementation could be improved, I would also prefer to have one single path to these steps instead of two. Not even speaking that the refactoring would be unrelated to scanner v4. So, unless you're foreseeing a simplification, I would instead simply be extending [EnrichWithVulnerability](https://github.com/stackrox/stackrox/blob/5563011163502bc87cd6bb953d2a51c6fee8a94b/pkg/images/enricher/enricher.go#L116) to accept the index report content and change things there to use the scanner v4 integration.